### DISCUSSION TITLE
I am not able to receive Attribute update

### DISCUSSION THREAD

**Discussion by markicos** [zigpy/zigpy#689] [Q&A] [answered]
Hi,

I have been working for some weeks with zigpy but i am still unable to get information from a temperature sensor, I join the device, I see the cluster are there in debug time, but i am not reaching `def attribute_updated(self, device, cluster, attribute_id, value)` in the main listener class.

I tried of adding listeners to the cluster when initializing a device
`class MainListener:
    """
    Contains callbacks that zigpy will call whenever something happens.
    Look for `listener_event` in the Zigpy source or just look at the logged warnings.
    """

    def __init__(self, application):
        self.application = application

    def device_initialized(self, device, *, new=True):
        """
        Called at runtime after a device's information has been queried.
        I also call it on startup to load existing devices from the DB.
        """
        LOGGER.info("Device is ready: new=%s, device=%s", new, device)

        for ep_id, endpoint in device.endpoints.items():
            # Ignore ZDO
            if ep_id == 0:
                continue

            # You need to attach a listener to every cluster to receive events
            for cluster in endpoint.in_clusters.values():
                # The context listener passes its own object as the first argument
                # to the callback
                cluster.add_context_listener(self)`

I don't see any frame in the serial bus with the request command, I don't understand how zigpy works I have no idea of how to sent a request. ¿any one can helpme for this.

------------------------------------------------------------------------------------------------------------------------------------

Later on i would like to add that i am trying to add a custom cluster with some measurement, this cluster has the ID=0xFFEE as i understood that it can be uses for manufacturer customs attributes, I implemented this cluster in a zigbee sensor and when it joins the nw i can see the cluster, but it is included in the `app.devices[ieee].end_points[10].out_clusters` instead of `in_cluster`, anyone knows why the cluster is in the out?

I defined also de new cluster in zigpy\zcl\clusters\manufacturer_specific.py as something like:

```
class xxx(Cluster):
    cluster_id = 0xFFEE
    name = "xxx"
    ep_attribute = "xxx"
    attributes = {
        # Air Thales Measurement Sensor
        0x0000: ("x_value", t.int16s),
        0x0001: ("y_value", t.int16s),

    }
    server_commands = {}
    client_commands = {}
```

**Comment by puddly** [comment, marked as answer]
For you to receive attribute reports from a relatively compliant device (i.e. not Aqara, which sends them to `0x0000` on its own), you have to tell the device you want to receive them:

```Python
# Report when the temperature changes by at least two degrees, at most once per minute and at least once per hour
# It will be reported as degrees C, multiplied by 100 (so -10C will be reported as `-10000`)
await device.endpoints[your_ep_id].temperature.configure_reporting("measured_value", 60, 3600, 200)
```

It'd be easiest to do this within `device_initialized` if `new == True`, so that you have time to make the request before the device starts aggressively sleeping.

**Reply by puddly** [reply]
As for why your cluster is in `out_clusters` instead of `in_clusters`, this is sent by your device in its Simple Descriptor response. If your device says it's an output cluster, zigpy puts it in `out_clusters`.

**Reply by markicos** [reply]
You were right, it was a problem in the sensor, now it is solved and i can see the cluster in `in_clusters`.

thanks,

**Comment by Adminiuga** [comment]
Configure_reporting tells the device it needs to send attribute updates. to tell the device where to send the attr updates you have to bind the cluster.

**Comment by markicos** [comment]
Thank you very much for your reply,

I tried to add this line in the code in both in the `main` code and in the `device_initialized` see bellow the behavior:

In the `device_initialized` I've got an exception as this function is not an async

In the 'main':
It seems it is sending the frame but i don't get any answer. :'(

DEBUG:zigpy_deconz.zigbee.application:Sending Zigbee request with tsn 19 under 20 request id, data: b'0013060000002901000a00c800'
DEBUG:zigpy_deconz.api:Command Command.aps_data_request (28, 20, 0, <DeconzAddressEndpoint address_mode=2 address=0xA8E2 endpoint=10>, 260, 1026, 1, b'\x00\x13\x06\x00\x00\x00)\x01\x00\n\x00\xc8\x00', 2, 0)
DEBUG:zigpy_deconz.uart:Send: 0x12300023001c00140002e2a80a04010204010d000013060000002901000a00c8000200
DEBUG:zigpy_deconz.uart:Frame received: 0x123000090002002214
DEBUG:zigpy_deconz.api:APS data request response: [2, <DeviceState.APSDE_DATA_REQUEST_SLOTS_AVAILABLE|2: 34>, 20]
DEBUG:zigpy_deconz.uart:Frame received: 0x0e31000700a600
DEBUG:zigpy_deconz.api:Device state changed response: [<DeviceState.128|APSDE_DATA_REQUEST_SLOTS_AVAILABLE|APSDE_DATA_CONFIRM|2: 166>, 0]
DEBUG:zigpy_deconz.uart:Frame received: 0x1c32000c00050002e2a8ffe7
DEBUG:zigpy_deconz.api:Command Command.aps_data_confirm (0,)
DEBUG:zigpy_deconz.uart:Send: 0x04310007000000
DEBUG:zigpy_deconz.uart:Frame received: 0x04310013000c00221402e2a80a010000000000
DEBUG:zigpy_deconz.api:APS data confirm response for request with id 20: 00
DEBUG:zigpy_deconz.api:Request id: 0x14 'aps_data_confirm' for <DeconzAddressEndpoint address_mode=ADDRESS_MODE.NWK address=0xa8e2 endpoint=10>, status: 0x00
DEBUG:zigpy_deconz.uart:Frame received: 0x0e32000700aa00
DEBUG:zigpy_deconz.api:Device state changed response: [<DeviceState.128|APSDE_DATA_REQUEST_SLOTS_AVAILABLE|APSDE_DATA_INDICATION|2: 170>, 0]
DEBUG:zigpy_deconz.uart:Frame received: 0x1c33000c00050002e2a8ffe7
DEBUG:zigpy_deconz.api:Command Command.aps_data_indication (1, 1)
DEBUG:zigpy_deconz.uart:Send: 0x1732000800010001
DEBUG:zigpy_deconz.uart:Frame received: 0x17320022001b00220200000102e2a80a0401020404001813070000afff00db0500e7
DEBUG:zigpy_deconz.api:APS data indication response: [27, <DeviceState.APSDE_DATA_REQUEST_SLOTS_AVAILABLE|2: 34>, <DeconzAddress address_mode=ADDRESS_MODE.NWK address=0x0000>, 1, <DeconzAddress address_mode=ADDRESS_MODE.NWK address=0xa8e2>, 10, 260, 1026, b'\x18\x13\x07\x00', 0, 175, 255, 0, 219, 5, 0, -25]
DEBUG:zigpy.zcl:[0xa8e2:10:0x0402] ZCL deserialize: <ZCLHeader frame_control=<FrameControl frame_type=GLOBAL_COMMAND manufacturer_specific=False is_reply=True disable_default_response=True> manufacturer=None tsn=19 command_id=Command.Configure_Reporting_rsp>
DEBUG:zigpy_deconz.api:'aps_data_indication' response from <DeconzAddress address_mode=ADDRESS_MODE.NWK address=0xa8e2>, ep: 10, profile: 0x0104, cluster_id: 0x0402, data: b'18130700'
DEBUG:zigpy_deconz.uart:Frame received: 0x1c33000c00050002e2a8ffe7
DEBUG:zigpy_deconz.uart:Frame received: 0x1c34000c00050002e2a8ffe7
DEBUG:zigpy_deconz.uart:Frame received: 0x1c35000c00050002e2a8ffe7
..... non stop


Finally I got and answer by using `read_attributes` `configure_reporting `, I found it read atributes by looking at the code, i barely understand how it works, or what exactly is the 'temperature' inside the endpoint.

`await app.get_device(ieee=ieee).endpoints[epid].temperature.read_attributes(["measured_value"])`

It is possible the device is not having configure_reporting but I am not sure...

**Reply by Adminiuga** [reply]
It configured the attribute reporting successfully. Did you bind the cluster ? await endpoint.temperature.bind() ?

**Reply by markicos** [reply]
Nop, i didn't let me check

**Reply by markicos** [reply]
Thanks a lot 
Of course it works.... for me this is like magic, after the bind i receive the temperature in the period i request.

I will read the code to learn about bind(), is there  any place(doc or whatever) where i should've gone before making those questions? because I haven't any clue of using the configure_reporting() neither to perform bind() after it.

**Reply by Adminiuga** [reply]
ZCL specs would be a good place to start, where they talk about the "foundation" and global ZCL commands. And then figuring out how zigpy implemented it.

**Reply by markicos** [reply]
Thanks

**Comment by markicos** [comment]
Hello,

I am now trying to do the same `.configure_reporting` with a new custom cluster, but it doesn't work, Don't know where the problem is, any idea thanks in advance.
with this code:
`    await app.get_device(ieee=ieee).endpoints[enid].new_cluster.configure_reporting("attri_0",1,5,200)
    await app.get_device(ieee=ieee).endpoints[enid].new_cluster.bind()`

here you are the log i've got.
```
DEBUG:zigpy_deconz.uart:Frame received: 0x17320025001e0022020000010237b90a0401eeff07001813078d00000000afff599f0600ee
DEBUG:zigpy_deconz.api:APS data indication response: [30, <DeviceState.APSDE_DATA_REQUEST_SLOTS_AVAILABLE|2: 34>, <DeconzAddress address_mode=ADDRESS_MODE.NWK address=0x0000>, 1, <DeconzAddress address_mode=ADDRESS_MODE.NWK address=0xb937>, 10, 260, 65518, b'\x18\x13\x07\x8d\x00\x00\x00', 0, 175, 255, 89, 159, 6, 0, -18]
DEBUG:zigpy.zcl:[0xb937:10:0xffee] ZCL deserialize: <ZCLHeader frame_control=<FrameControl frame_type=GLOBAL_COMMAND manufacturer_specific=False is_reply=True disable_default_response=True> manufacturer=None tsn=19 command_id=Command.Configure_Reporting_rsp>
DEBUG:zigpy_deconz.api:'aps_data_indication' response from <DeconzAddress address_mode=ADDRESS_MODE.NWK address=0xb937>, ep: 10, profile: 0x0104, cluster_id: 0xffee, data: b'1813078d000000'
DEBUG:zigpy.device:[0xb937] Extending timeout for 0x15 request
DEBUG:zigpy_deconz.zigbee.application:Sending Zigbee request with tsn 21 under 22 request id, data: b'150e71188d1336cef40aeeff03d69306ffff2e210001'
DEBUG:zigpy_deconz.api:Command Command.aps_data_request (37, 22, 0, <DeconzAddressEndpoint address_mode=2 address=0xB937 endpoint=0>, 0, <ZDOCmd.Bind_req: 0x0021>, 0, b'\x15\x0eq\x18\x8d\x136\xce\xf4\n\xee\xff\x03\xd6\x93\x06\xff\xff.!\x00\x01', 2, 0)
DEBUG:zigpy_deconz.uart:Send: 0x1233002c00250016000237b90000002100001600150e71188d1336cef40aeeff03d69306ffff2e2100010200
DEBUG:zigpy_deconz.uart:Frame received: 0x123300090002002216
DEBUG:zigpy_deconz.api:APS data request response: [2, <DeviceState.APSDE_DATA_REQUEST_SLOTS_AVAILABLE|2: 34>, 22]
DEBUG:zigpy_deconz.uart:Frame received: 0x0e34000700a600
DEBUG:zigpy_deconz.api:Device state changed response: [<DeviceState.128|APSDE_DATA_REQUEST_SLOTS_AVAILABLE|APSDE_DATA_CONFIRM|2: 166>, 0]
DEBUG:zigpy_deconz.api:Command Command.aps_data_confirm (0,)
DEBUG:zigpy_deconz.uart:Send: 0x04340007000000
DEBUG:zigpy_deconz.uart:Frame received: 0x04340013000c0022160237b900000000000000
DEBUG:zigpy_deconz.api:APS data confirm response for request with id 22: 00
DEBUG:zigpy_deconz.uart:Frame received: 0x0e35000700aa00
DEBUG:zigpy_deconz.api:Device state changed response: [<DeviceState.128|APSDE_DATA_REQUEST_SLOTS_AVAILABLE|APSDE_DATA_INDICATION|2: 170>, 0]
DEBUG:zigpy_deconz.api:Request id: 0x16 'aps_data_confirm' for <DeconzAddressEndpoint address_mode=ADDRESS_MODE.NWK address=0xb937 endpoint=0>, status: 0x00
DEBUG:zigpy_deconz.api:Command Command.aps_data_indication (1, 1)
DEBUG:zigpy_deconz.uart:Send: 0x1735000800010001
DEBUG:zigpy_deconz.uart:Frame received: 0x1735002000190022020000000237b900000021800200150000afff599f0600ef
DEBUG:zigpy_deconz.api:APS data indication response: [25, <DeviceState.APSDE_DATA_REQUEST_SLOTS_AVAILABLE|2: 34>, <DeconzAddress address_mode=ADDRESS_MODE.NWK address=0x0000>, 0, <DeconzAddress address_mode=ADDRESS_MODE.NWK address=0xb937>, 0, 0, 32801, b'\x15\x00', 0, 175, 255, 89, 159, 6, 0, -17]
DEBUG:zigpy_deconz.api:'aps_data_indication' response from <DeconzAddress address_mode=ADDRESS_MODE.NWK address=0xb937>, ep: 0, profile: 0x0000, cluster_id: 0x8021, data: b'1500'
DEBUG:zigpy_deconz.uart:Frame received: 0x1c36000c0005000237b9ffef
DEBUG:zigpy_deconz.uart:Frame received: 0x1c37000c0005000237b9ffef
DEBUG:zigpy_deconz.uart:Frame received: 0x1c38000c0005000237b9fff0

```
and no ends...


On the other hand i could read attributes by using:

    while True:
        await app.get_device(ieee=ieee).endpoints[enid].new_cluster.read_attributes(["attribute_0","attribute_1"])
        await asyncio.sleep(10)

**Comment by markicos** [comment]
Hello all,

I am glad to say that finally I the problem was solved, there was a mismatching between the data type defined in the zigbee sensor cluster and the data type in defined in the python cluster, so when I was configuring the reporting the sensor was reaching an exception because of unespected type.

Thank you all for your support
